### PR TITLE
Allow colored messages inside of colored messages

### PR DESCRIPTION
**Pull request by permcody**
This PR properly handles and formats *most* colored messages inside of colored messages. This occurs for instance when a warning messages are printed from a multiapp. There are two keys to making this work:

 1.  We need to be able to rely upon `_console` existing everywhere. This PR does this by adding a new global symbol by that name. This makes it possible to use _console from the mooseWarning() macro among others.

2. We need to not stomp on multi-line color codes common in error and warning messages. Extra logic to format multi-line color codes with the existing `indentMessage()` utility was added.

### CODE REVIEW

**Comment by mangerij** [comment]
what

**Comment by permcody** [comment]
@idaholab/moose-team - Making a new global namespace _console member works pretty well and allows us to use _console _nearly_ everywhere with the better buffering and formatting control that we get from using a "Moose" object over using the raw output streams. 

I ran into one exception that can be worked around so this isn't perfect. There's a material in BISON that has a `mooseWarning()` in a static member function. This unfortunately causes a compile error since the scope lookup rules apparently try to use the member _console instead of falling back to the global _console even though the former isn't valid. There are several work arounds to this, the easiest in this case is to just make it a const member method instead. It's a utility method and is never called statically. I could also make it so we could pass another optional parameter to these macros to force using the global scope...

Pros: Nicely formatted messages with multiple colors on a single line (warnings from subapps)
Cons: Doesn't work seamlessly 100% as outlined above

**Comment by permcody** [comment]
screen shot 2017-01-12 at 8 08 47 am

File: framework/include/base/MooseError.h [hunk 1]
```diff
@@ -42,10 +42,10 @@
   {                                                                                 \
     std::ostringstream _error_oss_;                                                 \
     _error_oss_ << "\n\n"                                                           \
-                << (Moose::_color_console ? XTERM_RED : "")                         \
+                << COLOR_RED                                                        \
```
Comment (jwpeterson) [inline framework/include/base/MooseError.h]: Regardless of whether the new global gets merged, these `COLOR` changes should be.

File: framework/src/base/Moose.C [hunk 2]
```diff
@@ -1193,3 +1193,5 @@ bool _deprecated_is_error = false;
 bool _throw_on_error = false;
 
 } // namespace Moose
+
+OStreamProxy & _console = Moose::out;
```
Comment (jwpeterson) [inline framework/src/base/Moose.C]: I need to research the portability of this more.  Can you really create a *static* reference to what is effectively `std::cout`?  The initialization order of static variables is undefined, so at the time this reference is created, `std::cout` might not exist?

Comment (jwpeterson) [inline framework/src/base/Moose.C]: This also seems like it should emit a bunch of shadowing warnings (two things named `_console` in the same scope) but we just don't have those warnings enabled.

**Review by jwpeterson — CHANGES_REQUESTED** [review]
Need to be 100% sure you can even have a static reference to an object like `std::cout`.

File: framework/include/base/MooseError.h (hunk 1 above)
Comment (permcody) [inline framework/include/base/MooseError.h]: Fortunately I broke that out in a separate commit.

File: framework/src/base/Moose.C (hunk 2 above)
Comment (jwpeterson) [inline framework/src/base/Moose.C]: I should not have used the word static, this is a *global* variable.  The [rule](http://stackoverflow.com/questions/3746238/c-global-initialization-order-ignores-dependencies/3746249#3746249) appears to be that global variables in a single translation unit are initialized in the order in which they are defined.  Assuming that `std::cout` is also a global variable of some kind, this should be fine.  Still not a big fan of the shadowing, however.

Comment (jwpeterson) [inline framework/src/base/Moose.C]: This is another good answer saying [it's guaranteed](http://stackoverflow.com/questions/8784892/is-stdcout-guaranteed-to-be-initialized) to work for C++11 and worked in practice for all C++03 compilers.

**Review by friedmud — APPROVED** [review]
This looks good to me.

Let's fixup BISON and we should add some documentation somewhere about when you can/can't use `mooseError()`/`mooseWarning()` to keep that from happening again.  In those cases we will just use `Moose::out` and `Moose::err`.  They should be very few and far between and aren't worth engineering an enormous solution around.

**Comment by permcody** [comment]
Sure, I'll document in `MooseError.h` the current restrictions.

File: framework/src/base/Moose.C (hunk 2 above)
Comment (permcody) [inline framework/src/base/Moose.C]: 👍

**Review by jwpeterson — APPROVED** [review]

**Comment by permcody** [comment]
This passes all tests!

**Pull request merged**